Silence unnecessary "missing `dyn`" errors and tweak E0746 suggestions

estebank opened the pull request:
Rework impl Trait and Box<dyn Trait> return type suggestions in E0746

* If the return value types diverge, mention it and do not suggest `impl Trait`
* If the returned `dyn Trait` is not object safe, mention it and do not suggest `Box<dyn Trait>` and boxing the returned values
* If the function has arguments with borrowed types, suggest `&dyn Trait`
* Tweak wording of `E0782`

```
error[E0746]: return type cannot have an unboxed trait object
  --> $DIR/dyn-trait-return-should-be-impl-trait.rs:13:13
   |
LL | fn bap() -> Trait { Struct }
   |             ^^^^^ doesn't have a size known at compile-time
   |
help: return an `impl Trait` instead of a `dyn Trait`
   |
LL | fn bap() -> impl Trait { Struct }
   |             ++++
help: alternatively, box the return type to make a boxed trait object, and wrap all of the returned values in `Box::new`
   |
LL | fn bap() -> Box<dyn Trait> { Box::new(Struct) }
   |             +++++++      +   +++++++++      +
```
```
error[E0746]: return type cannot have an unboxed trait object
  --> $DIR/object-unsafe-trait-in-return-position-dyn-trait.rs:36:18
   |
LL | fn can(x: &A) -> dyn NotObjectSafe {
   |                  ^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
   |
help: trait `NotObjectSafe` is not object safe, so you can't return a boxed trait object `Box<dyn NotObjectSafe>`
  --> $DIR/object-unsafe-trait-in-return-position-dyn-trait.rs:36:22
   |
LL | fn can(x: &A) -> dyn NotObjectSafe {
   |                      ^^^^^^^^^^^^^
help: return an `impl Trait` instead of a `dyn Trait`
   |
LL | fn can(x: &A) -> impl NotObjectSafe {
   |                  ~~~~
help: alternatively, you might be able to borrow from the function's argument
   |
LL | fn can(x: &A) -> &dyn NotObjectSafe {
   |                  +
```
```
error[E0782]: trait objects must include the `dyn` keyword
  --> $DIR/not-on-bare-trait-2021.rs:12:19
   |
LL | fn bar(x: Foo) -> Foo {
   |                   ^^^
   |
help: use `impl Foo` to return an opaque type, as long as you return a single underlying type
   |
LL | fn bar(x: Foo) -> impl Foo {
   |                   ++++
help: alternatively, you can return a boxed trait object
   |
LL | fn bar(x: Foo) -> Box<dyn Foo> {
   |                   +++++++    +
```

Silence "missing dyn" error if "?Sized type" error is already emitted.

Account for `type Alias = dyn Trait;` in unsized return suggestion:

```
error[E0746]: return type cannot have an unboxed trait object
 --> tests/ui/unsized/issue-91801.rs:8:77
  |
8 | fn or<'a>(first: &'static Validator<'a>, second: &'static Validator<'a>) -> Validator<'a> {
  |                                                                             ^^^^^^^^^^^^^ doesn't have a size known at compile-time
-Ztrack-diagnostics: created at compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:566:39
  |
help: box the return type to make a boxed trait object, and wrap all of the returned values in `Box::new`
  |
8 | fn or<'a>(first: &'static Validator<'a>, second: &'static Validator<'a>) -> Box<Validator<'a>> {
  |                                                                             ++++             +
help: alternatively, you might be able to borrow from one of the function's arguments
  |
8 | fn or<'a>(first: &'static Validator<'a>, second: &'static Validator<'a>) -> &Validator<'a> {
  |                                                                             +
```

Use more complex logic for E0277 deduplication of `FullfillmentError`s, accounting and customizing the span for `SizedArgument` obligations and functions with unsized return types. We only emit one unsized local error whenever possible.

Point at the let binding type or init expression instead of the pattern when possible for `?Sized` errors.

```
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
  --> $DIR/unsized-locals-using-unsized-fn-params.rs:13:15
   |
LL |     let _foo: [u8] = *foo;
   |               ^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `[u8]`
   = note: all local variables must have a statically known size
```

Fix #121037. Fix #105753.

estebank committed:
Rework `impl Trait` and `Box<dyn Trait>` return type suggestions in E0746

* If the return value types diverge, mention it and do not suggest `impl Trait`
* If the returned `dyn Trait` is not object safe, mention it and do not suggest `Box<dyn Trait>` and boxing the returned values
* If the function has arguments with borrowed types, suggest `&dyn Trait`
* Tweak wording of `E0782`

```
error[E0746]: return type cannot have an unboxed trait object
  --> $DIR/dyn-trait-return-should-be-impl-trait.rs:13:13
   |
LL | fn bap() -> Trait { Struct }
   |             ^^^^^ doesn't have a size known at compile-time
   |
help: return an `impl Trait` instead of a `dyn Trait`
   |
LL | fn bap() -> impl Trait { Struct }
   |             ++++
help: alternatively, box the return type to make a boxed trait object, and wrap all of the returned values in `Box::new`
   |
LL | fn bap() -> Box<dyn Trait> { Box::new(Struct) }
   |             +++++++      +   +++++++++      +
```
```
error[E0746]: return type cannot have an unboxed trait object
  --> $DIR/object-unsafe-trait-in-return-position-dyn-trait.rs:36:18
   |
LL | fn can(x: &A) -> dyn NotObjectSafe {
   |                  ^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
   |
help: trait `NotObjectSafe` is not object safe, so you can't return a boxed trait object `Box<dyn NotObjectSafe>`
  --> $DIR/object-unsafe-trait-in-return-position-dyn-trait.rs:36:22
   |
LL | fn can(x: &A) -> dyn NotObjectSafe {
   |                      ^^^^^^^^^^^^^
help: return an `impl Trait` instead of a `dyn Trait`
   |
LL | fn can(x: &A) -> impl NotObjectSafe {
   |                  ~~~~
help: alternatively, you might be able to borrow from the function's argument
   |
LL | fn can(x: &A) -> &dyn NotObjectSafe {
   |                  +
```
```
error[E0782]: trait objects must include the `dyn` keyword
  --> $DIR/not-on-bare-trait-2021.rs:12:19
   |
LL | fn bar(x: Foo) -> Foo {
   |                   ^^^
   |
help: use `impl Foo` to return an opaque type, as long as you return a single underlying type
   |
LL | fn bar(x: Foo) -> impl Foo {
   |                   ++++
help: alternatively, you can return a boxed trait object
   |
LL | fn bar(x: Foo) -> Box<dyn Foo> {
   |                   +++++++    +
```

estebank committed:
Account for `type Alias = dyn Trait;` in unsized return suggestion

```
error[E0746]: return type cannot have an unboxed trait object
 --> tests/ui/unsized/issue-91801.rs:8:77
  |
8 | fn or<'a>(first: &'static Validator<'a>, second: &'static Validator<'a>) -> Validator<'a> {
  |                                                                             ^^^^^^^^^^^^^ doesn't have a size known at compile-time
-Ztrack-diagnostics: created at compiler/rustc_trait_selection/src/traits/error_reporting/type_err_ctxt_ext.rs:566:39
  |
help: box the return type to make a boxed trait object, and wrap all of the returned values in `Box::new`
  |
8 | fn or<'a>(first: &'static Validator<'a>, second: &'static Validator<'a>) -> Box<Validator<'a>> {
  |                                                                             ++++             +
help: alternatively, you might be able to borrow from one of the function's arguments
  |
8 | fn or<'a>(first: &'static Validator<'a>, second: &'static Validator<'a>) -> &Validator<'a> {
  |                                                                             +
```

estebank committed:
Centralize calculation of FullfillmentError span

Use a method for caulculating deduplication and presentation span.

estebank committed:
On implicit `Sized` bound on fn argument, point at type instead of pattern

estebank committed:
Tweak `Sized` obligation spans on fn argument `impl Trait`

Remove now unnecessary span label.

estebank committed:
Silence "missing `dyn`" error if "`!Sized` type" error is already emitted

matthiaskrgr commented on src/tools/clippy/clippy_lints/src/missing_asserts_for_indexing.rs:
```diff
@@ -10,9 +10,14 @@ use rustc_ast::{LitKind, RangeLimits};
 use rustc_data_structures::packed::Pu128;
 use rustc_data_structures::unhash::UnhashMap;
 use rustc_errors::{Applicability, Diag};
+<<<<<<< Updated upstream
 use rustc_hir::{BinOp, Block, Body, Expr, ExprKind, UnOp};
+=======
+use rustc_hir::{BinOp, Block, Body, Expr, ExprKind, StmtKind, UnOp};
+>>>>>>> Stashed changes
 use rustc_lint::{LateContext, LateLintPass};
```
?

estebank replied:
Just noticed it, removing it just now.

matthiaskrgr commented on src/tools/clippy/clippy_lints/src/missing_asserts_for_indexing.rs:
```diff
@@ -136,9 +141,28 @@ fn assert_len_expr<'hir>(
 ) -> Option<(LengthComparison, usize, &'hir Expr<'hir>)> {
     if let Some(higher::If { cond, then, .. }) = higher::If::hir(expr)
         && let ExprKind::Unary(UnOp::Not, condition) = &cond.kind
+<<<<<<< Updated upstream
         && let ExprKind::Binary(bin_op, left, right) = &condition.kind
+=======
+        && let ExprKind::Block(block, None) = condition.kind
+        && let [local] = block.stmts
+        && let StmtKind::Let(local) = local.kind
+        && let Some(condition) = local.init
+        && let Some((bin_op, left, right)) = {
+            let expn = condition.span.ctxt().outer_expn_data();
+            match condition.kind {
+                ExprKind::Binary(bin_op, left, right)
+                if expn.edition >= Edition::Edition2024 => Some((bin_op, left, right)),
+                ExprKind::Unary(UnOp::Not, expr)
+                if let ExprKind::Unary(UnOp::Not, expr) = expr.kind
+                    && let ExprKind::Binary(bin_op, left, right) = expr.kind
+                => Some((bin_op, left, right)),
+                _ => None,
+            }
+        }
+>>>>>>> Stashed changes
 
-        && let Some((cmp, asserted_len, slice_len)) = len_comparison(*bin_op, left, right)
+        && let Some((cmp, asserted_len, slice_len)) = len_comparison(bin_op, left, right)
```
some more here

estebank replied:
Removed

estebank committed:
Remove unnecessary `info!`

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8403530251/job/23014129447) [(plain)](https://github.com/rust-lang/rust/commit/f60ded9135c892c59968e54a72fc84c10a3bba39/checks/23014129447/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
    Checking if_chain v1.0.2
error[E0308]: mismatched types
   --> src/tools/clippy/clippy_lints/src/missing_asserts_for_indexing.rs:141:70
    |
141 |         && let Some((cmp, asserted_len, slice_len)) = len_comparison(bin_op, left, right)
    |                                                       -------------- ^^^^^^ expected `Spanned<BinOpKind>`, found `&Spanned<BinOpKind>`
    |                                                       arguments to this function are incorrect
    |
    = note: expected struct `rustc_span::source_map::Spanned<_>`
            found reference `&rustc_span::source_map::Spanned<_>`
            found reference `&rustc_span::source_map::Spanned<_>`
note: function defined here
   --> src/tools/clippy/clippy_lints/src/missing_asserts_for_indexing.rs:100:4
    |
100 | fn len_comparison<'hir>(
101 |     bin_op: BinOp,
    |     -------------
help: consider dereferencing the borrow
    |
    |
141 |         && let Some((cmp, asserted_len, slice_len)) = len_comparison(*bin_op, left, right)

For more information about this error, try `rustc --explain E0308`.
error: could not compile `clippy_lints` (lib) due to 2 previous errors
Build completed unsuccessfully in 0:02:54
```

</details>

estebank committed:
On `fn` with no borrowed argument, explain that `&dyn Trait` could have been an option

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8403618267/job/23014314966) [(plain)](https://github.com/rust-lang/rust/commit/77af1661ce686be99103d84a9499f868c693a8bb/checks/23014314966/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain

Location:
    /cargo/registry/src/index.crates.io-6f17d22bba15001f/ui_test-0.22.2/src/lib.rs:395:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
command: CLIPPY_CONF_DIR="tests" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-7dc543e856f87725.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-f8a01377ac0ff203.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-0f3bb1d43666e40c.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-dd131ad522aed9e0.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-ae846d6788d6a9f1.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-9152caa60d22e5d6.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-389517e109d775ca.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-93d7303c6012a33b.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-2d43560082bb21c6.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-969c152412fe4d84.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps/libserde_derive-b0de0647ab33f326.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-f944c6c97a7094c6.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-eddc5a1e301140d8.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--crate-type=lib" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui/crashes" "tests/ui/crashes/ice-6251.rs" "--edition" "2021"
error: actual output differed from expected
Execute `cargo uibless` to update `tests/ui/crashes/ice-6251.stderr` to the actual output
--- tests/ui/crashes/ice-6251.stderr
error: test failed, to rerun pass `--test compile-test`
error: test failed, to rerun pass `--test compile-test`
+++ <stderr output>
 error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
-  --> tests/ui/crashes/ice-6251.rs:4:45
+  --> tests/ui/crashes/ice-6251.rs:4:48
    |
 LL | fn bug<T>() -> impl Iterator<Item = [(); { |x: [u8]| x }]> {
+   |                                                ^^^^ doesn't have a size known at compile-time
    |
    = help: the trait `std::marker::Sized` is not implemented for `[u8]`
... 26 lines skipped ...
---
full stderr:
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
##[error]  --> tests/ui/crashes/ice-6251.rs:4:48
   |
LL | fn bug<T>() -> impl Iterator<Item = [(); { |x: [u8]| x }]> {
   |
   = help: the trait `std::marker::Sized` is not implemented for `[u8]`
   = help: unsized fn params are gated as an unstable feature
help: function arguments must have a statically known size, borrowed slices always have a known size
help: function arguments must have a statically known size, borrowed slices always have a known size
   |
LL | fn bug<T>() -> impl Iterator<Item = [(); { |x: &[u8]| x }]> {

error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
##[error]  --> tests/ui/crashes/ice-6251.rs:4:54
   |
   |
LL | fn bug<T>() -> impl Iterator<Item = [(); { |x: [u8]| x }]> {
   |
   = help: the trait `std::marker::Sized` is not implemented for `[u8]`
   = note: the return type of a function must have a statically known size


error[E0308]: mismatched types
##[error]  --> tests/ui/crashes/ice-6251.rs:4:44
   |
LL | fn bug<T>() -> impl Iterator<Item = [(); { |x: [u8]| x }]> {
   |
   = note: expected type `usize`
   = note: expected type `usize`
           found closure `{closure@tests/ui/crashes/ice-6251.rs:4:44: 4:53}`
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
```

</details>

estebank committed:
fix clippy test

fee1-dead commented:
I don't think I have the capacity to review this right now.

r? compiler

compiler-errors commented on compiler/rustc_infer/src/traits/mod.rs:
```diff
@@ -206,6 +206,50 @@ impl<'tcx> FulfillmentError<'tcx> {
         FulfillmentError { obligation, code, root_obligation }
     }
+
+    pub fn span(&self, tcx: TyCtxt<'tcx>) -> Span {
```
I feel like it's incredibly easy to forget to use this function over `.obligation.cause.span` - can you consider another solution that doesn't give devs more choices that cause subtle inconsistencies when used differently?

estebank replied:
The issue about forgetting to call the method is fair, but do notice that we're already duplicating logic to modify that span after the fact in the current code. I'm concerned about properly tracking the span from inception if that will require additional memory or have non-trivial runtime impact on non-diagnostic codepaths. I'll see what options we have available to us that don't have those problems.

estebank replied:
I had to still use a closure for the deduplication logic (as it is very local to this one E0277 error for `Sized` locals), but otherwise the obligation span is now the one we want for all errors.

estebank committed:
On fn argument obligation, point at type instead of pattern

estebank committed:
Remove `FullfillmentError::span` method

estebank committed:
Remove unused argument

estebank committed:
Deduplicate `?Sized` binding errors to only their declaration, not use

* Point at let binding type or init that is `?Sized`, as appropriate
* On `?Sized` binding *access*, do not emit an error as we will already have emitted one on the binding

estebank committed:
Add test for calling fn with unsized return type

estebank committed:
Account for `-> ?Sized` fn calls when deduplicating unsized local errors

estebank committed:
Drive-by formatting change: import types used from `hir` instead of `hir::`

estebank committed:
Further dedup unsized locals errors involving method calls

estebank committed:
Fix rebase: `Node::Local` rename

estebank commented:
Apologies for dumping more commits to the same PR. Each commit is locally independent (they could be their own PRs), but each individual step is building a tiny bit more on the previous one. I believe that there are may be a handful of additional things that could be done here (particularly, reduce redundancy between E0038 and E0277), but likely nothing more that should be added to this one. If the load is too much, feel free to review as many commits in a row as you feel comfortable, and I'll split the rest to their own PR(s).

estebank commented:
Blocker: this needs to compile-pass

```
error[E0277]: the size for values of type `str` cannot be known at compilation time
 --> f108.rs:2:16
  |
2 |     let ref x: str = *"";
  |                ^^^ doesn't have a size known at compile-time
  |
  = help: the trait `Sized` is not implemented for `str`
  = note: all local variables must have a statically known size
  = help: unsized locals are gated as an unstable feature
help: consider borrowing here
  |
2 |     let ref x: &str = *"";
  |                +
```

estebank committed:
Do not error on `let ref x: str = *"";`

estebank committed:
Account for tuples and structs in unsized locals errors for more accurate spans

```
error[E0277]: the size for values of type `X` cannot be known at compilation time
  --> $DIR/unsized6.rs:26:19
   |
LL | fn f3<X: ?Sized>(x1: Box<X>, x2: Box<X>, x3: Box<X>) {
   |       - this type parameter needs to be `Sized`
...
LL |     let (y, z) = (*x3, 4);
   |                   ^^^ doesn't have a size known at compile-time
   |
   = note: all local variables must have a statically known size
   = help: unsized locals are gated as an unstable feature
help: consider removing the `?Sized` bound to make the type parameter `Sized`
   |
LL - fn f3<X: ?Sized>(x1: Box<X>, x2: Box<X>, x3: Box<X>) {
LL + fn f3<X>(x1: Box<X>, x2: Box<X>, x3: Box<X>) {
   |
```
```
error[E0277]: the size for values of type `dyn T` cannot be known at compilation time
  --> $DIR/unsized-binding.rs:12:28
   |
LL |     let Foo(a, b) = Foo(1, bar());
   |                            ^^^^^ doesn't have a size known at compile-time
   |
   = help: the trait `Sized` is not implemented for `dyn T`
   = note: all local variables must have a statically known size
   = help: unsized locals are gated as an unstable feature
```

estebank commented on tests/ui/unsized-locals/suggest-borrow.stderr:
```diff
@@ -44,10 +44,10 @@ LL |     let x: &[u8] = &vec!(1, 2, 3)[..];
 
 error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
-  --> $DIR/suggest-borrow.rs:4:9
+  --> $DIR/suggest-borrow.rs:4:19
    |
 LL |     let x: [u8] = &vec!(1, 2, 3)[..];
-   |         ^ doesn't have a size known at compile-time
+   |                   ^^^^^^^^^^^^^^^^^^ doesn't have a size known at compile-time
```
This is weird: it should point at the type, not the init expr. If I modify the test to make this the only thing here, then *it does the right thing*. I'm not *too* concerned because of the close locality but it is technically incorrect (as the expr has type `&[u8]`, *the binding* is `[u8]`).

estebank commented on tests/ui/sized/unsized-binding.stderr:
```diff
@@ -1,13 +1,51 @@
    = help: unsized locals are gated as an unstable feature
 
-error: aborting due to 1 previous error
+error[E0277]: the size for values of type `dyn T` cannot be known at compilation time
+  --> $DIR/unsized-binding.rs:12:28
+   |
+LL |     let Foo(a, b) = Foo(1, bar());
+   |                            ^^^^^ doesn't have a size known at compile-time
+   |
+   = help: the trait `Sized` is not implemented for `dyn T`
+   = note: all local variables must have a statically known size
+   = help: unsized locals are gated as an unstable feature
+
+error[E0277]: the size for values of type `dyn T` cannot be known at compilation time
+  --> $DIR/unsized-binding.rs:12:28
+   |
+LL |     let Foo(a, b) = Foo(1, bar());
+   |                            ^^^^^ doesn't have a size known at compile-time
+   |
+   = help: the trait `Sized` is not implemented for `dyn T`
+   = note: all function arguments must have a statically known size
+   = help: unsized fn params are gated as an unstable feature
```
This two could be deduplicated if we reworded the last two note and help to have the same text.

estebank replied:
The last commit does this. In stable, these two errors will collapse into a single one.

estebank committed:
Emit single explanatory note on unsized locals and fn params instead of two

On stable, unify wording of unsized locals and unsized fn params so the deduplication machinery can trigger.

On nightly, mention the nightly feature to enable unsized locals or unsized fn params.

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8526639602/job/23356315734) [(plain)](https://github.com/rust-lang/rust/commit/fb70fea77038b285947cf79bb7895bbe709ac733/checks/23356315734/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
called `Result::unwrap()` on an `Err` value: tests failed

FAILED TEST: tests/ui/crashes/ice-6251.rs

command: CLIPPY_CONF_DIR="tests" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/clippy-driver" "--error-format=json" "--emit=metadata" "-Aunused" "-Ainternal_features" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Dwarnings" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps" "--extern=clippy_config=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_config-d901a432f9b5437d.rlib" "--extern=clippy_lints=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_lints-c39b683b0ce40234.rlib" "--extern=clippy_utils=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libclippy_utils-a841394a3899e7db.rlib" "--extern=futures=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libfutures-b9aa5555d3eeb00c.rlib" "--extern=if_chain=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libif_chain-ae846d6788d6a9f1.rlib" "--extern=itertools=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libitertools-9152caa60d22e5d6.rlib" "--extern=parking_lot=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libparking_lot-6014a4e4c8724228.rlib" "--extern=quote=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libquote-93d7303c6012a33b.rlib" "--extern=regex=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libregex-dd6285fa210e40f1.rlib" "--extern=serde=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libserde-f6a8ee4fbadda337.rlib" "--extern=serde_derive=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps/libserde_derive-7f0d708317b92523.so" "--extern=syn=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libsyn-5e386b163d38144b.rlib" "--extern=tokio=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/libtokio-b7759248f5057cd2.rlib" "-Ldependency=/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/release/deps" "--crate-type=lib" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/ui_test/tests/ui/crashes" "tests/ui/crashes/ice-6251.rs" "--edition" "2021"
    /cargo/registry/src/index.crates.io-6f17d22bba15001f/ui_test-0.22.2/src/lib.rs:395:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: test failed, to rerun pass `--test compile-test`

---
   --> tests/ui/crashes/ice-6251.rs:4:48
... 3 lines skipped ...
    |
    = help: the trait `std::marker::Sized` is not implemented for `[u8]`
-   = help: unsized fn params are gated as an unstable feature
+   = help: unsized fn params are gated as unstable feature `#[feature(unsized_fn_params)]`
 help: function arguments must have a statically known size, borrowed slices always have a known size
... 23 lines skipped ...
 Some errors have detailed explanations: E0277, E0308.
 For more information about an error, try `rustc --explain E0277`.



full stderr:
error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
##[error]  --> tests/ui/crashes/ice-6251.rs:4:48
   |
LL | fn bug<T>() -> impl Iterator<Item = [(); { |x: [u8]| x }]> {
   |
   = help: the trait `std::marker::Sized` is not implemented for `[u8]`
   = help: the trait `std::marker::Sized` is not implemented for `[u8]`
   = help: unsized fn params are gated as unstable feature `#[feature(unsized_fn_params)]`
help: function arguments must have a statically known size, borrowed slices always have a known size
   |
LL | fn bug<T>() -> impl Iterator<Item = [(); { |x: &[u8]| x }]> {

error[E0277]: the size for values of type `[u8]` cannot be known at compilation time
##[error]  --> tests/ui/crashes/ice-6251.rs:4:54
   |
   |
LL | fn bug<T>() -> impl Iterator<Item = [(); { |x: [u8]| x }]> {
   |
   = help: the trait `std::marker::Sized` is not implemented for `[u8]`
   = note: the return type of a function must have a statically known size


error[E0308]: mismatched types
##[error]  --> tests/ui/crashes/ice-6251.rs:4:44
   |
LL | fn bug<T>() -> impl Iterator<Item = [(); { |x: [u8]| x }]> {
   |
   = note: expected type `usize`
   = note: expected type `usize`
           found closure `{closure@tests/ui/crashes/ice-6251.rs:4:44: 4:53}`
error: aborting due to 3 previous errors

Some errors have detailed explanations: E0277, E0308.
For more information about an error, try `rustc --explain E0277`.
```

</details>

estebank commented:
r? @oli-obk 

This is the PR we talked about earlier today. You can review each commit independently and I'll split them out to their own PR.

oli-obk commented on compiler/rustc_trait_selection/src/traits/error_reporting/suggestions.rs:
```diff
@@ -1835,11 +1899,77 @@ impl<'tcx> TypeErrCtxt<'_, 'tcx> {
-            sugg,
-            Applicability::MaybeIncorrect,
-        );
+        let object_unsafe: Vec<_> = traits
```
The function is called `suggest_impl_trait`, but this doesn't do that. Needs a name update and doc update.

Please split up this function into three functions (`impl Trait`, `Box<dyn Trait>`, and `&dyn Trait`).

alex-semenyuk commented:
@estebank
Ping from triage. Could you please post your status on this PR? This PR has not been updated in more than 15 days.
It has merge conflicts and not addressed comment

fmease commented:
They are on vacation rn and will return soon

JohnCSimon commented:
> They are on vacation rn and will return soon

@estebank - Triage - Pinging again after few months, do you have a status update? Thanks.
